File: android/cast/src/main/java/com/dreampany/cast/vm/MoreViewModel.java
package com.dreampany.cast.vm;

import android.app.Activity;
import android.app.Application;

import com.dreampany.cast.data.model.More;
import com.dreampany.cast.ui.enums.MoreType;
import com.dreampany.cast.ui.model.MoreItem;
import com.dreampany.cast.ui.model.UiTask;
import com.dreampany.framework.data.model.Response;
import com.dreampany.framework.misc.AppExecutors;
import com.dreampany.framework.misc.ResponseMapper;
import com.dreampany.framework.misc.RxMapper;
import com.dreampany.framework.util.SettingsUtil;
import com.dreampany.framework.ui.vm.BaseViewModel;
import com.dreampany.network.manager.NetworkManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import javax.inject.Inject;

import io.reactivex.Maybe;
import io.reactivex.disposables.Disposable;

/**
 * Created by Hawladar Roman on 7/24/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class MoreViewModel extends BaseViewModel<More, MoreItem, UiTask<More>> {

    @Inject
    MoreViewModel(Application application,
                  RxMapper rx,
                  AppExecutors ex,
                  ResponseMapper rm,
                  NetworkManager network) {
        super(application, rx, ex, rm);
    }


    public void loads(boolean important) {
        if (!takeAction(important, getMultipleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(getItems())
                .doOnSubscribe(subscription -> postProgress(true))
                .subscribe(result -> postResult(Response.Type.ADD, result, true), this::postFailures);
        addMultipleSubscription(disposable);
    }

    private Maybe<List<MoreItem>> getItems() {
        return Maybe.fromCallable(() -> {
            List<MoreItem> items = new ArrayList<>();
            items.add(MoreItem.getItem(new More(MoreType.SETTINGS.name(), 0, MoreType.SETTINGS)));
            items.add(MoreItem.getItem(new More(MoreType.APPS.name(), 0, MoreType.APPS)));
            items.add(MoreItem.getItem(new More(MoreType.RATE_US.name(), 0 , MoreType.RATE_US)));
            items.add(MoreItem.getItem(new More(MoreType.FEEDBACK.name(), 0, MoreType.FEEDBACK)));
            //items.add(MoreItem.getItemBySymbolRx(new More(MoreType.INVITE)));
            items.add(MoreItem.getItem(new More(MoreType.LICENSE.name(), 0, MoreType.LICENSE)));
            items.add(MoreItem.getItem(new More(MoreType.ABOUT.name(), 0, MoreType.ABOUT)));
            return items;
        });
    }

    public void moreApps(Activity activity) {
        SettingsUtil.moreApps(Objects.requireNonNull(activity));
    }

    public void rateUs(Activity activity) {
        SettingsUtil.rateUs(activity);
    }

    public void sendFeedback(Activity activity) {
        SettingsUtil.feedback(activity);
    }
}

File: android/nearby/src/main/java/com/dreampany/nearby/ui/more/vm/MoreViewModel.kt
package com.dreampany.nearby.ui.more.vm

import android.app.Application
import com.dreampany.framework.misc.func.ResponseMapper
import com.dreampany.framework.misc.func.SmartError
import com.dreampany.framework.ui.model.UiTask
import com.dreampany.framework.ui.vm.BaseViewModel
import com.dreampany.theme.Colors
import com.dreampany.nearby.R
import com.dreampany.nearby.data.enums.Action
import com.dreampany.nearby.data.enums.State
import com.dreampany.nearby.data.enums.Subtype
import com.dreampany.nearby.data.enums.Type
import com.dreampany.nearby.data.model.more.More
import com.dreampany.nearby.ui.more.model.MoreItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

/**
 * Created by roman on 21/3/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class MoreViewModel
@Inject constructor(
    application: Application,
    rm: ResponseMapper,
    private val colors: Colors
) : BaseViewModel<Type, Subtype, State, Action, More, MoreItem, UiTask<Type, Subtype, State, Action, More>>(
    application,
    rm
) {

    fun loadMores() {
        uiScope.launch {
            //postProgress(true)
            val result = getMores()
            postResult(result.toItems())
        }
    }

    private suspend fun getMores() = withContext(Dispatchers.IO) {
        val result = arrayListOf<More>()
        /*result.add(
            More(
                Type.MORE, Subtype.SETTINGS, R.drawable.ic_crypto, R.string.title_feature_crypto,
                colors.nextColor(Type.MORE.name)
            )
        )*/
        result.add(
            More(
                Type.MORE, Subtype.APPS, R.drawable.ic_apps_black_24dp, R.string.more_apps,
                colors.nextColor(Type.MORE.name)
            )
        )
        result.add(
            More(
                Type.MORE, Subtype.RATE_US, R.drawable.ic_rate_review_black_24dp, R.string.rate_us,
                colors.nextColor(Type.MORE.name)
            )
        )
        result.add(
            More(
                Type.MORE, Subtype.FEEDBACK, R.drawable.ic_feedback_black_24dp, R.string.title_feedback,
                colors.nextColor(Type.MORE.name)
            )
        )
        result.add(
            More(
                Type.MORE, Subtype.ABOUT, R.drawable.ic_info_black_24dp, R.string.about,
                colors.nextColor(Type.MORE.name)
            )
        )
        /*result.add(
            More(
                Type.MORE, Subtype.LICENSE, R.drawable.ic_security_black_24dp, R.string.license,
                colors.nextColor(Type.MORE.name)
            )
        )*/
        /*result.add(
            More(
                Type.MORE, Subtype.ABOUT, R.drawable.ic_info_black_24dp, R.string.about,
                colors.nextColor(Type.MORE.name)
            )
        )*/
        result
    }

    private suspend fun List<More>.toItems(): List<MoreItem> {
        val list = this
        return withContext(Dispatchers.IO) {
            list.map { MoreItem(it) }
        }
    }

    private fun postProgress(progress: Boolean) {
        postProgressMultiple(
            Type.MORE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            progress = progress
        )
    }


    private fun postError(error: SmartError) {
        postMultiple(
            Type.MORE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            error = error,
            showProgress = false
        )
    }

    private fun postResult(result: List<MoreItem>) {
        postMultiple(
            Type.MORE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            result = result,
            showProgress = false
        )
    }

}